sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2385,20 +2392,63 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
       // `CalendarInterval` doesn't have enough info. For instance, new CalendarInterval(0, 0, 0)
       // can be derived from INTERVAL '0-0' YEAR TO MONTH as well as from
       // INTERVAL '0 00:00:00' DAY TO SECOND.
+      val fromUnit =
+        ctx.errorCapturingUnitToUnitInterval.body.from.getText.toLowerCase(Locale.ROOT)
       val toUnit = ctx.errorCapturingUnitToUnitInterval.body.to.getText.toLowerCase(Locale.ROOT)
       if (toUnit == "month") {
         assert(calendarInterval.days == 0 && calendarInterval.microseconds == 0)
-        // TODO(SPARK-35773): Parse year-month interval literals to tightest types
-        Literal(calendarInterval.months, YearMonthIntervalType())
+        val start = YearMonthIntervalType.stringToField(fromUnit)
+        Literal(calendarInterval.months, YearMonthIntervalType(start, YearMonthIntervalType.MONTH))
       } else {
         assert(calendarInterval.months == 0)
-        val fromUnit =
-          ctx.errorCapturingUnitToUnitInterval.body.from.getText.toLowerCase(Locale.ROOT)
         val micros = IntervalUtils.getDuration(calendarInterval, TimeUnit.MICROSECONDS)
         val start = DayTimeIntervalType.stringToField(fromUnit)
         val end = DayTimeIntervalType.stringToField(toUnit)
         Literal(micros, DayTimeIntervalType(start, end))
       }
+    } else if (ctx.errorCapturingMultiUnitsInterval != null && !conf.legacyIntervalEnabled) {
+      val units =
+        ctx.errorCapturingMultiUnitsInterval.body.unit.asScala.map(
+          _.getText.toLowerCase(Locale.ROOT).stripSuffix("s"))
+      val yearMonthFields = Set.empty[Byte]
+      val dayTimeFields = Set.empty[Byte]
+      for (unit <- units) {
+        if (YearMonthIntervalType.stringToField.contains(unit)) {
+          yearMonthFields += YearMonthIntervalType.stringToField(unit)
+        } else if (DayTimeIntervalType.stringToField.contains(unit)) {
+          dayTimeFields += DayTimeIntervalType.stringToField(unit)
+        } else if (unit == "week") {
+          dayTimeFields += DayTimeIntervalType.DAY
+        } else {
+          assert(unit == "millisecond" || unit == "microsecond")
+          dayTimeFields += DayTimeIntervalType.SECOND
+        }
+      }
+      if (yearMonthFields.nonEmpty) {
+        if (dayTimeFields.nonEmpty) {
+          val matchIter =
+            multiUnitsPattern.findAllMatchIn(ctx.errorCapturingMultiUnitsInterval.body.getText)
+          val literalBuilder = new StringBuilder
+          matchIter.foreach { matched =>
+            literalBuilder.append(matched.group(1) + " ")
+            literalBuilder.append(matched.group(2))
+            if (matched.group(3) != null) {
+              // In case a unit ends with "s"
+              literalBuilder.append(matched.group(3))
+            }
+            if (matchIter.hasNext) {
+              literalBuilder.append(" ")
+            }
+          }
+          throw QueryParsingErrors.mixedIntervalUnitsError(literalBuilder.toString, ctx)
```

Comment:
how about
```
val multiUnitsContext = ctx.errorCapturingMultiUnitsInterval.body
val values = multiUnitsContext.intervalValue().asScala.map(_.getText)
val units = multiUnitsContext.unit.asScala.map(_.getText)
val literalStr = values.zip(units).map { case (v, u) => v + " " + u }.mkString(" ")
throw ...
```

Reply:
It's almost the same done in `visitMultiUnitsInterval` isn't it?
As I mentioned [here](https://github.com/apache/spark/pull/32949#discussion_r666850179), I wanted to avoid the building logic twice for literal string.

Reply:
It saves the checking logic: https://github.com/apache/spark/pull/32949/commits/94ade45232987732fac9e53c36c3c51ceb859fa9#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0faL2471-L2477

Reply:
OK I think we must duplicate some code: either the validation code or the string building code. 

Perf is not an issue here as it's for the error case, and I found the string builder code is simpler and OK to duplicate.

Reply:
If it's too much trouble, let's don't put the interval literal string in the error message. The error contains the line number and start position, people can see the original interval literal string by themselves.

Reply:
BTW did you try `source(ctx.errorCapturingMultiUnitsInterval.body)`? In CTAS we use `source` to get the original SQL text of the query

Reply:
> BTW did you try source(ctx.errorCapturingMultiUnitsInterval.body)? In CTAS we use source to get the original SQL text of the query

Oh, I didn't notice that. I'll try it and if it seems work, let's use it. Thank you.

Reply:
`source` seems to work. Thank you!